### PR TITLE
(1/x) Break out models and api, refactor classes

### PR DESCRIPTION
**Pull request by atrakh**
This is the start of a series of PRs that will break ld-vscode into more manageable pieces.

- Refactor existing classes into singletons
- Break out API code and data types into separate classes / files
- Rename `LDFlagManager` to `FlagStore` and move most unrelated functionality out into separate functions.

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import * as rp from 'request-promise-native';
+import * as url from 'url';
+
+import { configuration } from './configuration';
+import { Flag } from './models';
+
+// LaunchDarklyAPI is a wrapper around request-promise-native for requesting data from LaunchDarkly's REST API. The caller is expected to catch all exceptions.
```
Comment (atrakh) [inline src/api.ts]: This should eventually be replaced by one of our fancy api clients, but just setting this up as scaffolding to make other work easier. Behavior is very similar to the old `getFeatureFlag` function in `flags.ts`

File: src/configuration.ts [hunk 2]
```diff
@@ -1,59 +1,17 @@
...
-	/**
-	 * The LaunchDarkly base uri to be used. Optional.
-	 */
-	baseUri: string;
-
-	/**
-	 * The LaunchDarkly stream uri to be used. Optional.
-	 */
-	streamUri: string;
-}
```
Comment (atrakh) [inline src/configuration.ts]: We no longer pass around configuration settings, instead, all packages read from the singleton exported in this file.

File: src/configuration.ts [hunk 3]
```diff
@@ -1,59 +1,17 @@
...
 	constructor() {
 		this.reload();
 	}
 
 	reload() {
-		let config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration('launchdarkly');
+		let config: WorkspaceConfiguration = workspace.getConfiguration('launchdarkly');
 		for (const option in this) {
-			this[option] = config[option];
+			this[option] = config.get(option);
```
Comment (atrakh) [inline src/configuration.ts]: Fixes a typescript warning

File: src/flags.ts [hunk 4]
```diff
@@ -102,64 +72,75 @@ export function isPrecedingCharStringDelimeter(document: vscode.TextDocument, po
...
-		}
 
-		this.updateProcessor = StreamProcessor(settings.sdkKey, config, Requestor(settings.sdkKey, config));
+	constructor() {
+		this.store = InMemoryFeatureStore();
 		this.start();
 	}
 
+	reload(e: ConfigurationChangeEvent) {
+		if (['sdkKey', 'baseUri', 'streamUri'].some(option => e.affectsConfiguration(`launchdarkly.${option}`))) {
```
Comment (atrakh) [inline src/flags.ts]: Cleaner logic for checking if we need to reload the store, which also gives us the benefit of not having to pass around configuration settings

Old method is down here: https://github.com/launchdarkly/ld-vscode/pull/26/files#diff-53649f4ba3d51e10eff913927cc17e0cL144

File: src/flags.ts [hunk 5]
```diff
@@ -102,64 +72,75 @@ export function isPrecedingCharStringDelimeter(document: vscode.TextDocument, po
...
-						errMsg = `Error retrieving feature flags: ${err.message}.`;
-					} else {
-						console.error(err);
-						errMsg = `Unexpected error retrieving flags.`;
-					}
-					vscode.window.showErrorMessage(`[LaunchDarkly] ${errMsg}`);
+		if (!config.sdkKey || !config.baseUri || !config.streamUri) {
+			console.warn('LaunchDarkly extension is not configured. Language support is unavailable.');
+			return;
+		}
```
Comment (atrakh) [inline src/flags.ts]: Moved this logic into `start`. It's slightly different in that we're checking for all the crucial values for streaming to work.

File: src/flags.ts [hunk 6]
```diff
@@ -170,92 +151,105 @@ export class LDFlagManager implements IFlagManager {
 		};
 	}
 
-	registerProviders(ctx: vscode.ExtensionContext, settings: IConfiguration) {
+	registerProviders(ctx: ExtensionContext) {
```
Comment (atrakh) [inline src/flags.ts]: This method shouldn't be part of the `FlagStore`. I'm going to factor it out in another PR.

File: src/models.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	environments: Map<String, Environment>;
+}
+
+export class Environment extends Resource {
+	_site: { href: string };
+}
+
+export class FlagConfiguration {
+	key: string;
+	variations: Array<any>;
```
Comment (ldhenry) [inline src/models.ts]: It'd be nice if we could get more specific types here. I have a feeling we can import this type from https://github.com/launchdarkly/api-client-typescript but I haven't tried it yet

Comment (atrakh) [reply src/models.ts]: I was hoping to use api-client-typescript for both the API and types, but unfortunately we haven't released it as an npm package yet. I'll ask around to see if we can get that out.
